tokio/src/runtime/coop.rs

```diff
@@ -240,6 +243,42 @@ cfg_coop! {
             self.0.is_none()
         }
     }
+
+    pin_project! {
+        /// A future type that calls `poll_proceed` before polling the inner future to check if the
+        /// inner future has exceeded its budget. If the inner future resolves, this will
+        /// automatically call `RestoreOnPending::made_progress` before resolving this future with
+        /// the result of the inner one. If polling the inner future is pending, polling this future
+        /// type will also return a `Poll::Pending`.
+        #[must_use = "futures do nothing unless polled"]
+        pub(crate) struct BudgetConstraintFuture<F: Future> {
+            #[pin]
+            pub(crate) fut: F,
+        }
+    }
+
+    impl<F: Future> Future for BudgetConstraintFuture<F> {
+        type Output = F::Output;
+
+        fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+            let coop = ready!(poll_proceed(cx));
+            let me = self.project();
+            if let Poll::Ready(ret) = me.fut.poll(cx) {
+                coop.made_progress();
+                Poll::Ready(ret)
+            } else {
+                Poll::Pending
+            }
+        }
+    }
+
+    /// Run a future with a budget constraint for cooperative scheduling.
+    /// If the future exceeds its budget while being polled, control is yielded back to the
+    /// runtime.
+    #[inline]
+    pub(crate) async fn budget_constraint<F: Future>(fut: F) -> F::Output {
```

Comment:
How about calling this `with_budget`?

Reply:
There is already a free function called `with_budget` in that namespace so I would suggest maybe `run_with_budget` as it runs the given future in some sense?

Reply:
Maybe just define the struct like this:
```
struct Coop<F>(F);
```
Then you can construct it as `Coop(my_fut).await`.

Reply:
Hmm, we could do this but I like it better to have a free function here that takes a future and can be awaited similar to `tokio::task::Unconstrained` future that is created by `tokio::task::unconstrained` function. But since it's only a private future your suggestion would also be ok I think.

Naming is hard :)

Reply:
Sure, makes sense to me.